pkg/util/helper/status_test.go

```diff
@@ -0,0 +1,273 @@
+			expectedOp:    controllerutil.OperationResultNone,
+			expectedError: "Internal error occurred: status update failed",
+		},
+		{
+			name: "no changes needed",
+			setupObj: &corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test-pod",
+					Namespace: "default",
+				},
+				Status: corev1.PodStatus{
+					Phase: corev1.PodRunning,
+				},
+			},
+			mutateStatus: func(_ *corev1.Pod) {
+				// No changes to status
+			},
+			statusError:   false,
+			expectedOp:    controllerutil.OperationResultNone,
+			expectedError: "",
+		},
+		{
+			name:     "object not found",
+			setupObj: nil, // Not create the object
+			mutateStatus: func(pod *corev1.Pod) {
+				pod.Status.Phase = corev1.PodRunning
+			},
+			statusError:   false,
+			expectedOp:    controllerutil.OperationResultNone,
+			expectedError: "not found",
+		},
+		{
+			name: "mutation error",
+			setupObj: &corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test-pod",
+					Namespace: "default",
+				},
+			},
+			mutateStatus: func(pod *corev1.Pod) {
+				// Simulate mutation error by changing name
+				pod.Name = "different-name"
+			},
+			statusError:   false,
+			expectedOp:    controllerutil.OperationResultNone,
+			expectedError: "cannot mutate object name",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			clientBuilder := fake.NewClientBuilder().WithScheme(scheme)
+			if tt.setupObj != nil {
+				clientBuilder = clientBuilder.WithObjects(tt.setupObj)
+			}
+			fakeClient := clientBuilder.Build()
+
+			var client client.Client
+			if tt.statusError {
+				client = &mockClient{
+					Client:      fakeClient,
+					shouldError: true,
+				}
+			} else {
+				client = fakeClient
+			}
+
+			// Create a new object for update
+			obj := &corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test-pod",
+					Namespace: "default",
+				},
+			}
+
+			// Run the update
+			op, err := UpdateStatus(context.TODO(), client, obj, func() error {
+				if tt.mutateStatus != nil {
+					tt.mutateStatus(obj)
+				}
+				return nil
+			})
+
+			// Check error
+			if tt.expectedError != "" {
+				assert.Error(t, err)
+				assert.Contains(t, err.Error(), tt.expectedError)
+			} else {
+				assert.NoError(t, err)
+			}
+
+			// Check operation result
+			assert.Equal(t, tt.expectedOp, op)
+
+			// If successful update, verify the status was actually changed
+			if tt.expectedOp == controllerutil.OperationResultUpdatedStatusOnly {
+				updatedPod := &corev1.Pod{}
+				err := client.Get(context.TODO(), types.NamespacedName{Name: "test-pod", Namespace: "default"}, updatedPod)
+				assert.NoError(t, err)
+				assert.Equal(t, corev1.PodRunning, updatedPod.Status.Phase)
+			}
+		})
+	}
+}
+
+func TestMutate(t *testing.T) {
+	tests := []struct {
+		name          string
+		key           types.NamespacedName
+		obj           *corev1.Pod
+		mutateFn      controllerutil.MutateFn
+		expectedError string
+	}{
+		{
```

Comment:
Can you add test cases for `mutating object name/namespace`? Although `func TestUpdateStatus` does cover related scenarios, as a unit test, individual test functions should aim to measure as comprehensive a range of scenarios as possible

Reply:
Yes, i'll add it.